cpp/data_lang.cc

```diff
@@ -165,20 +165,18 @@ BigStr* ShellEncodeString(BigStr* s, int ysh_fallback) {
 namespace pyj8 {
 
 bool PartIsUtf8(BigStr* s, int start, int end) {
-  uint32_t codepoint;
-  uint32_t state = UTF8_ACCEPT;
-
-  for (int i = start; i < end; ++i) {
-    // This var or a static_cast<> is necessary.  Should really change BigStr*
-    // to use unsigned type
-    uint8_t c = s->data_[i];
-    decode(&state, &codepoint, c);
-    if (state == UTF8_REJECT) {
+  Utf8Result result;
+
+  for (int i = start; i < end;) {
+    utf8_decode(reinterpret_cast<unsigned char*>(&s->data_[i]), &result);
```

Comment:
All the reinterpret_cast<> in our code bugs me, though I checked CPython and it the same issue.

They do casting whenever they pick a byte out of a string

The char vs. unsigned char thing is such a wart in C

----

I think this would be slightly cleaner as `s->data_ + i`

It's basically staying in "pointer land" rather than getting a value and taking its address

Also the precedence of `&` and `->` and `[]` confuses me so I like that style